### DISCUSSION TITLE
Add Support for FreeRTOS for OpenAMP RPU Apps

### DISCUSSION THREAD

**Discussion by bentheredonethat** [OpenAMP/open-amp#592] [Ideas]
Today OpenAMP RPU apps have hard-coded helper.c and example main functions that assume they are for baremetal OS.

Below is proposal for design to add Xilinx-AMD supported FreeRTOS additions:

1. Move GIC Init logic that today is always assumed for baremetal to its own file called gic_init.c and pick up either baremetal or freertos gic_init.c. Link against either gic_init.c::app_gic_init() based on the target OS in CMake. See below diagram for changes.

![img1](https://github.com/OpenAMP/open-amp/assets/6487229/c5538387-9b20-480b-abce-a561265a5510)

2. OS-specific main routine - For each of the  demos add OS specific files with the main routine to link against. Link against apps/example/<demo>/<OS>/main.c::main() routine based on target OS. See below diagram:

![img2](https://github.com/OpenAMP/open-amp/assets/6487229/068c0b7a-da02-4896-9dbd-d4c3647fb4fa)

**Comment by bentheredonethat** [comment]
Apologies for formatting of the diagrams... 

- white rectangle is existing dir
- white round edged rectangle is existing source file
- yellow rectangle is new dir
- yellow round edged rectangle is new  source file

CC @arnopo  @tnmysh @wmamills

**Comment by arnopo** [comment]
Hi @bentheredonethat 
Before discussing such evolution the priority would be to move apps in https://github.com/OpenAMP/openamp-system-reference
do you discuss the topic internally with @tnmysh ?
